Group function definitions

vedant1811 opened the pull request:
Fixes #5405 

<!--### Your checklist for this pull request

Please review the [guidelines for contributing](http://solidity.readthedocs.io/en/latest/contributing.html) to this repository.

Please also note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.
-->

### Description

<!--
Please explain the changes you made here.

Thank you for your help!
-->

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

stackenbotten commented:
There was an error when running `test_check_style` for commit `4154f8cc043a71e73f8a8f5b85a8504377758310`:
```
Coding style error:
 libyul/backends/evm/EVMCodeTransform.cpp:834: const FunctionDefinition *seriesStart;
 libyul/backends/evm/EVMCodeTransform.cpp:837: const FunctionDefinition *functionDefinition = boost::get<FunctionDefinition>(&statement);

```
Please check that your changes are working as intended.

chriseth commented on libyul/AsmData.h:
```diff
@@ -70,7 +70,20 @@ struct VariableDeclaration { langutil::SourceLocation location; TypedNameList va
+	TypedNameList returnVariables;
+	Block body;
+
+	/// jump "over" to a specific statement
```
Please don't extend the AST just for the purpose of code generation. This information should be stored inside EVMCodeTransform.

chriseth commented:
@vedant1811 do you still want to continue on this? Do you need some help? Please don't hesitate to ask in https://gitter.im/ethereum/solidity-dev

stackenbotten commented:
There was an error when running `test_check_style` for commit `156435ee9e0f1a214afceb56c85b953478da5fec`:
```
Error: Trailing whitespace found:
 libyul/backends/evm/EVMCodeTransform.cpp:887: 

```
Please check that your changes are working as intended.

Commit:
[Yul] Directly jump over a series of function definitions

Implement a AbstractAssembly::setStackHeight function
Update the tests
Update Changelog

stackenbotten commented:
There was an error when running `test_check_style` for commit `3e604e402616ed2d6cf329ec1a1518808b457038`:
```
Coding style error:
 libyul/backends/evm/EVMCodeTransform.cpp:727: const AbstractAssembly::LabelID* seriesStart = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:731: const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);
 libyul/backends/evm/EVMCodeTransform.cpp:736: const AbstractAssembly::LabelID labelId = m_assembly.newLabelId();

```
Please check that your changes are working as intended.

vedant1811 commented:
The custom test case generated the following expected output:

```
Pretty printed source:
object "Contract" {
    code {
        {
            function f()
            {
            }
            function g()
            {
            }
            sstore(0, 1)
        }
    }
}


Binary representation:
600f566008565b565b600e565b565b5b6001600055

Text representation:
  jump(tag_1)
    /* "test.yul":41:56   */
  jump(tag_2)
tag_3:
    /* "test.yul":54:56   */
  jump
tag_2:
    /* "test.yul":57:72   */
  jump(tag_4)
tag_5:
    /* "test.yul":70:72   */
  jump
tag_4:
tag_1:
    /* "test.yul":83:84   */
  0x01
    /* "test.yul":80:81   */
  0x00
    /* "test.yul":73:85   */
  sstore
```

I need to fix 1 test failure, and add tests for the above optimisation

stackenbotten commented:
There was an error when running `test_check_style` for commit `6965b7ffaeacfdc08feb1c9abb0f03f88a76df71`:
```
Coding style error:
 libyul/backends/evm/EVMCodeTransform.cpp:729: const AbstractAssembly::LabelID* seriesLabel = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:730: const Statement* seriesEnd = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:850:const std::unordered_map<Statement const*, Statement const*> CodeTransform::groupFunctionDefinitions(vector<Statement> const& _statements) const
 libyul/backends/evm/EVMCodeTransform.cpp:854: const Statement* seriesStart = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:855: const Statement* lastFd = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:859: const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);
 libyul/backends/evm/EVMCodeTransform.h:226: const std::unordered_map<Statement const*, Statement const*> groupFunctionDefinitions(std::vector<Statement> const& _statements) const;

```
Please check that your changes are working as intended.

vedant1811 commented:
> There was an error when running `test_check_style` for commit `6965b7ffaeacfdc08feb1c9abb0f03f88a76df71`:
> 
> ```
> Coding style error:
>  libyul/backends/evm/EVMCodeTransform.cpp:729: const AbstractAssembly::LabelID* seriesLabel = nullptr;
>  libyul/backends/evm/EVMCodeTransform.cpp:730: const Statement* seriesEnd = nullptr;
>  libyul/backends/evm/EVMCodeTransform.cpp:850:const std::unordered_map<Statement const*, Statement const*> CodeTransform::groupFunctionDefinitions(vector<Statement> const& _statements) const
>  libyul/backends/evm/EVMCodeTransform.cpp:854: const Statement* seriesStart = nullptr;
>  libyul/backends/evm/EVMCodeTransform.cpp:855: const Statement* lastFd = nullptr;
>  libyul/backends/evm/EVMCodeTransform.cpp:859: const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);
>  libyul/backends/evm/EVMCodeTransform.h:226: const std::unordered_map<Statement const*, Statement const*> groupFunctionDefinitions(std::vector<Statement> const& _statements) const;
> ```
> 
> Please check that your changes are working as intended.

Error statements are not very helpful. I am guessing I shouldn't be writing ` = nullptr;` in variable definitions but I'm not sure what's wrong with `const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);`

stackenbotten commented:
There was an error when running `test_check_style` for commit `a4cae840fbd201e2e7826183a06d787076fcee22`:
```
Coding style error:
 libyul/backends/evm/EVMCodeTransform.cpp:729: const AbstractAssembly::LabelID* seriesLabel = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:730: const Statement* seriesEnd = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:850:const std::unordered_map<Statement const*, Statement const*> CodeTransform::groupFunctionDefinitions(vector<Statement> const& _statements) const
 libyul/backends/evm/EVMCodeTransform.cpp:854: const Statement* seriesStart = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:855: const Statement* lastFd = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:859: const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);
 libyul/backends/evm/EVMCodeTransform.h:226: const std::unordered_map<Statement const*, Statement const*> groupFunctionDefinitions(std::vector<Statement> const& _statements) const;

```
Please check that your changes are working as intended.

vedant1811 commented:
I have no idea what's failing here: https://circleci.com/gh/ethereum/solidity/130187?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link :(

chriseth commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,10 +724,30 @@ void CodeTransform::visitExpression(Expression const& _expression)
+			auto label = m_assembly.newLabelId();
+			m_assembly.appendJumpTo(label, 0);
+
+			seriesLabel = &label;
```
I think the error is here, you are storing the address of a stack variable and then it goes out of scope.

chriseth commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,10 +724,30 @@ void CodeTransform::visitExpression(Expression const& _expression)
 
 void CodeTransform::visitStatements(vector<Statement> const& _statements)
 {
+	auto fdSeries = groupFunctionDefinitions(_statements);
```
Do you really need such a complicated analysis beforehand?
What about something like this:
```
void CodeTransform::visitStatements(vector<Statement> const& _statements)
{
  boost::optional<AbstractAssembly::LabelID> jumpTarget;
  for (auto const& statement: _statements)
  {
    if (statement is a function definition && !jumpTarget)
    {
      jumpTarget = m_assembly.newLabelId();
      m_assembly.appendJumpTo(*jumpTarget, 0);
    }
    if (statement is not a function definition && jumpTarget)
    {
      m_assembly.appendLabel(*jumpTarget)
      jumpTarget = boost::none;
    }
    freeUnusedVariables();
    boost::apply_visitor(*this, statement);
  }
  freeUnusedVariables();
  if jumpTarget)
    m_assembly.appendLabel(*jumpTarget)
}
```

vedant1811 replied:
Yes the analysis is needed. We don't want to add the `JUMP` statements around single functions. So something like this is needed:

			// create mapping only for series longer than 1
			if (seriesLength > 1)
				fdSeries[seriesStart] = lastFd;

chriseth replied:
Why not add it around single functions? If you always add it here, you can remove it from the code generation of FunctionDefinition.

vedant1811 replied:
Conditionally removing it from code generation of FunctionDefinition, caused the StackHeight errors (as we had discussed on chat). Unconditionally removing them will break logic for non-single functions.

chriseth replied:
Why does it break the logic for non-single functions?

vedant1811 replied:
Yes you're right, it won't break for non-single functions.

But I just tested removing the code from FunctionDefinition and the `stack height mismatch` errors are back

stackenbotten commented:
There was an error when running `test_check_style` for commit `d30dab020e5184d9fae4349eeb83d9451580c9ff`:
```
Coding style error:
 libyul/backends/evm/EVMCodeTransform.cpp:730: const Statement* seriesEnd = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:849:const std::unordered_map<Statement const*, Statement const*> CodeTransform::groupFunctionDefinitions(vector<Statement> const& _statements) const
 libyul/backends/evm/EVMCodeTransform.cpp:853: const Statement* seriesStart = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:854: const Statement* lastFd = nullptr;
 libyul/backends/evm/EVMCodeTransform.cpp:858: const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);
 libyul/backends/evm/EVMCodeTransform.h:226: const std::unordered_map<Statement const*, Statement const*> groupFunctionDefinitions(std::vector<Statement> const& _statements) const;

```
Please check that your changes are working as intended.

chriseth commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,8 +720,21 @@ void CodeTransform::visitExpression(Expression const& _expression)
+			jumpTarget = boost::none;
+		}
+
 		freeUnusedVariables();
```
I think it is better to do this first.

chriseth commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -489,16 +489,13 @@ void CodeTransform::operator()(FunctionDefinition const& _function)
 	m_assembly.setSourceLocation(_function.location);
 	int stackHeightBefore = m_assembly.stackHeight();
-	AbstractAssembly::LabelID afterFunction = m_assembly.newLabelId();
 
```
It might be best to implement a `setStackHeight` function in `AbstractAssembly` and call it here with `m_assembly.setStackHeight(height)`. At the end of the function, call it again with `m_assembly.setStackHeight(stackHeightBefore)`.

vedant1811 replied:
The function will just set `m_stackHeight` in `EVMAssembly` (and other derived classes)?

I can do that, but I have no idea what is happening 😅

ekpyron replied:
I think I got it now - ``setStackHeight`` has to be called *after* the ``if`` in the next line - in the ``m_evm15 == true`` case ``appendBeginSub`` will itself adjust the stack for the function arguments - but that's also done with ``height``, so currently that's done twice. So just moving ``m_assembly.setStackHeight(height)`` one statement down past the ``if`` should be all that's needed.

ekpyron replied:
Note that this is exactly the difference between ``-stackHeightBefore`` and ``-stackHeightBefore + height`` we had before this PR in the ``if(m_evm15)``.

vedant1811 commented:
@chriseth I'm free from my last job, so I can try to overlap 1-2 hours of work with your timings.

This PR is >1 month old; I want to finish it ASAP :)

vedant1811 commented:
I am still getting `Stack height mismatch between analysis and code generation phase` in 2 test cases.

ekpyron commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -589,8 +587,8 @@ void CodeTransform::operator()(FunctionDefinition const& _function)
 	m_stackAdjustment -= localStackAdjustment;
-	m_assembly.appendLabel(afterFunction);
 	checkStackHeight(&_function);
+	m_assembly.setStackHeight(stackHeightBefore);
```
What happens if you move this one line up?

vedant1811 replied:
Nothing. I get the same 2 test failures.

vedant1811 commented:
Nothing. I get the same 2 test failures.

On Fri, Jun 14, 2019 at 1:32 AM Daniel Kirchner <notifications@github.com>
wrote:

> *@ekpyron* commented on this pull request.
> ------------------------------
>
> In libyul/backends/evm/EVMCodeTransform.cpp
> <https://github.com/ethereum/solidity/pull/6666#discussion_r293494570>:
>
> >  	checkStackHeight(&_function);
> +	m_assembly.setStackHeight(stackHeightBefore);
>
> What happens if you move this one line up?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/solidity/pull/6666?email_source=notifications&email_token=AAY37SZX362H6HYRMM7LUO3P2KABJA5CNFSM4HK53XE2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3PQUGQ#pullrequestreview-249498138>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAY37S4PRLWQ5YCR4Y6W57DP2KABJANCNFSM4HK53XEQ>
> .
>

ekpyron commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,11 +722,29 @@ void CodeTransform::visitExpression(Expression const& _expression)
 
 void CodeTransform::visitStatements(vector<Statement> const& _statements)
 {
+	boost::optional<AbstractAssembly::LabelID> jumpTarget = boost::make_optional(false, AbstractAssembly::LabelID());
```
```suggestion
	boost::optional<AbstractAssembly::LabelID> jumpTarget;
```

That's not the stack problem, but still :-). (The default constructor will create an uninitialized ``optional`` as well)

vedant1811 replied:
There's a boost bug; I added a link to it in a comment

ekpyron replied:
Hm... the link refers to GCC < 5.1... we recently increased the minimal version to 5.0... and we require a minimum of boost 1.65 (whereas the link is for boost 1.63), but it's in boost 1.65 as well it seems (https://www.boost.org/doc/libs/1_65_0/libs/optional/doc/html/boost_optional/tutorial/gotchas/false_positive_with__wmaybe_uninitialized.html) - so strictly speaking, I guess we might still be affected by this...

My guess would be that we do this at other places in the code as well... but yeah - I'd say that both keeping this workaround and removing it is fine, so as you like ;-)!

vedant1811 replied:
@ekpyron Yeah I got a unexpected CI failure so I had to put this workaround.
Could you resolve the request changes so that this PR can be merged?

ekpyron commented:
Note: the failures is for EVM 1.5 cases.

ekpyron commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -488,19 +488,18 @@ void CodeTransform::operator()(FunctionDefinition const& _function)
 	m_assembly.setSourceLocation(_function.location);
-	int stackHeightBefore = m_assembly.stackHeight();
-	AbstractAssembly::LabelID afterFunction = m_assembly.newLabelId();
+	const int stackHeightBefore = m_assembly.stackHeight();
```
```suggestion
	int const stackHeightBefore = m_assembly.stackHeight();
```

ekpyron commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -488,19 +488,18 @@ void CodeTransform::operator()(FunctionDefinition const& _function)
+	const int stackHeightBefore = m_assembly.stackHeight();
 
 	if (m_evm15)
 	{
```
No brackets if it's a single line - the same for the ``else`` below.

ekpyron commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,11 +723,32 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	{
 		freeUnusedVariables();
+
+		const FunctionDefinition* functionDefinition = boost::get<FunctionDefinition>(&statement);
```
```suggestion
		FunctionDefinition const* functionDefinition = boost::get<FunctionDefinition>(&statement);
```

vedant1811 replied:
Should I just use `auto functionDefinition ...`?

ekpyron replied:
Fine with me!

ekpyron reviewed (commented):
Added some coding-style related comments - it'd be great, if you could address those :-)!

ekpyron reviewed (commented):
The tests are passing and this looks good to me now! But maybe best, if @chriseth takes another look - he should be back from vacation next week.

chriseth commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,11 +720,32 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	{
 		freeUnusedVariables();
+
+		auto functionDefinition = boost::get<FunctionDefinition>(&statement);
```
```suggestion
		auto const* functionDefinition = boost::get<FunctionDefinition>(&statement);
```

chriseth commented:
Looks good! Please rebase and squash into a single commit and add a changelog entry.

vedant1811 commented:
@chriseth, @ekpyron Looks like the relevant CI tests have passed and we can merge this PR.

chriseth commented on Changelog.md:
```diff
@@ -25,6 +25,7 @@ Compiler Features:
  * Yul: Adds break and continue keywords to for-loop syntax.
  * Yul: Support ``.`` as part of identifiers.
  * Yul Optimizer: Adds steps for detecting and removing of dead code.
+ * Yul Optimizer: Directly jump over a series of function definitions (instead of jumping over each one)
```
```suggestion
 * Yul Code Generator: Directly jump over a series of function definitions (instead of jumping over each one)
```

chriseth commented on libyul/backends/evm/EVMCodeTransform.cpp:
```diff
@@ -724,11 +720,32 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	for (auto const& statement: _statements)
 	{
 		freeUnusedVariables();
+
```
I think it would be better to add a `m_assembly.setSourceLocation(locationOf(statement));` here.

vedant1811 replied:
Shouldn't that go inside the `if (functionDefinition && !jumpTarget)` block; We only want to set the source location if its the start of a series of function definitions, right?

chriseth replied:
Either you put it before `appendJumpTo` and before `appendLabel` or here - it should be re-set inside the apply_visitor anyway.
